examples/browser-service-worker/src/service.js

```diff
@@ -0,0 +1,530 @@
+
+
+/**
+ * Fetches content from the IPFS and responds with it. 
+ * 
+ * @param {Object} options
+ * @param {Fetch} options.event
+ * @param {string} options.path
+ */
+const fetchContent = async ({ event, path }) => {
+  const [,protocol] = path.split('/')
+  switch (protocol) {
+    case 'ipns': {
+      return await fetchIPNSContent({
+        event,
+        path
+      })
+    }
+    case 'ipfs':
+      return await fetchIPFSContent({
+        event,
+        path
+     })
+    default: {
+      const response = await unsupportedProtocol(protocol)
+      return response
+    }
+  }
+}
+
+/**
+ * @param {Object} options
+ * @param {Fetch} options.event 
+ * @param {string} options.path
+ */
+const fetchIPNSContent = async ({/* path, event */}) => {
+  // Challenge: Implement IPNS support.
+  return new Response(`<html>
+  <body>
+    <h1>IPNS protocol support is not implemented</h1>
+    <p>It is left as an excercise to the viewer</p>
+  </body>
+</html>`, {
+    statusText: "IPNS support is not implemented",
+    status: 502
+  })
+}
+
+
+/**
+ * @param {Object} options
+ * @param {Fetch} options.event 
+ * @param {string} options.path
+ */
+const fetchIPFSContent = async ({ event, path }) => {
+  // Obtains IPFS instance
+  const ipfs = await new IPFSService().use(event)
+  try {
+    // Attempt to fetch the file for the path.
+    return await fetchIPFSFile(ipfs, path)
+  } catch ({ message }) {
+    // Above can fail because path might be for a directory, if so
+    // respond with diretory listing.
+    if (message.includes('dag node is a directory')) {
+      // Ensure ends with `/` so that generated file URLs would resolve properly
+      if (path.endsWith('/')) {
+        return fetchIPFSDirectory(ipfs, path)
+      } else {
+        return Response.redirect(`${event.request.url}/`)
+      }
+    }
+
+    // If such link does not exists respond with 404
+    if (message.startsWith('no link named')) {
+      return new Response(message, {
+        statusText: message,
+        status: 404
+      })
+    }
+
+    // If problem with CID respond with 400
+    if (message.startsWith('multihash length inconsistent') || 
+        message.startsWith('Non-base58 character'))
+    {
+      return new Response(message, {
+        statusText: message,
+        status: 400
+      })
+    }
+
+    // Otherwise respond with 500
+    return new Response(message, {
+      statusText: message,
+      status: 500
+    })
+  }
+}
+
+/**
+ * @param {IPFS} ipfs 
+ * @param {string} path
+ */
+const fetchIPFSFile = async (ipfs, path) => {
+  const content = ipfs.cat(path)
+  // We get back `AsyncIterable<Uint8Array>`, but if such file does not exists
+  // we will only know once we start reading. To responding with a stream that
+  // fails we attempt to read first chunk and only then produce a response
+  // instace. This ensures that we either error or return Response that doesn't
+  // immediately fail.
+  const chunks = content[Symbol.asyncIterator]()
```

Comment:
Could you go into a bit more detail about the problem this is trying to solve?

Just because the first file chunk was received, there's no guarantee the rest of the chunks are available so how does getting the first chunk before continuing help here?

Reply:
I'm not sure how to make this comment any more clear, could you please suggest something ?

So if you look at `fetchIPFSContent` function you can see that it tries to use `fetchIPFSFile` first and if it fails due to `dag node is a directory` it falls back to `fetchIPFSDirectory` (that renders diretory index) or if failed with `no link named` it returns 404.

Without pulling first chunk these errors happens inside the stream, meaning request will succeed with 200, but than it's body will fail and nothing is served.

That is why I mentioned other day that APIs that just return `AsyncIterable`s aren't very ergonomic and can easily lead to mistakes, because they tangle two different class of errors:

1. "incorrect requests"
2. "io error satisfying requests"

I think they are fundamentally different because they require very different recovery logic.

Reply:
If you want to know if a CID resolves to a file or a directory before catting it, you can use `ipfs.files.stat` and make a decision based on that.

It'll also tell you if the file exists (by which I mean blocks are present in the local repo, I'm assuming that's what the comment means too, please correct if not), sort of, in that if the CID cannot be resolved it'll try to pull it from the network and will eventually time out.

My point is that if a file consists of 100 chunks, just because the first chunk arrives there's no guarantee that the 68th chunk will arrive so the stream can still fail, though from your clarification above I don't think that's the problem you are trying to solve.

Reply:
@Gozala I think this is the last thing to be resolved before it can be merged. Unless there's a good reason to leave it as it is can you please refactor to use `ipfs.files.stat` to find out what type of DAG is being accessed instead of using exceptions for flow control?

Reply:
So motivation here was to demonstrate some of the nitty gritty details of the API. I suspect that any serious usage of SW will try to avoid additional roundtrips for every single request (which in common cases would be redundant). And I feel like it is easy to overlook problem so showing how to address that was deliberate.

If you do think it is out of scope for this example, I can go ahead and change it as suggested.

Reply:
I think maybe this is a bit more advanced than is necessary for an example and having it be concise and easy to digest would be more valuable.

It's a fair point though, how about adding a 'further reading' or 'what next' section to the readme with the sort of problems you're likely to encounter and how to resolve them?